### PR TITLE
A huge client refactor

### PR DESCRIPTION
**Pull request by jasoncalabrese**
Goal is to get rid of the code smell in client.js

### CODE REVIEW

File: lib/client/index.js [hunk 1]
```diff
@@ -0,0 +1,1710 @@
...
+    }
+
+    // If this is delta, calculate the difference, merge and sort
+    var diff = nsArrayDiff(cachedDataArray,receivedDataArray);
+    return cachedDataArray.concat(diff).sort(function(a, b) {
+      return a.mills - b.mills;
+    });
+  }
+
+  socket.on('dataUpdate', function receivedSGV(d) {
```
Comment (jasoncalabrese) [inline lib/client/index.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (8)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799258271/issues/source?bid=2452379&fileBranchId=2725007#l1439)

File: lib/client/index.js [hunk 2]
```diff
@@ -0,0 +1,1710 @@
+'use strict';
+
+var $ = (global && global.$) || require('jquery');
+var _ = require('lodash');
+var d3 = (global && global.d3) || require('d3');
+
+var language = require('../language')();
+var profile = require('../profilefunctions')();
```
Comment (jasoncalabrese) [inline lib/client/index.js]: Codacy Issue found: ['profile' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799258271/issues/source?bid=2452379&fileBranchId=2725007#l8)

File: static/js/ui-utils.js [hunk 3]
```diff
@@ -3,119 +3,7 @@
 var openDraw = null;
 
 function rawBGsEnabled() {
```
Comment (jasoncalabrese) [inline static/js/ui-utils.js]: Codacy Issue found: ['rawBGsEnabled' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799258312/issues/source?bid=2452379&fileBranchId=2725007#l5)

File: lib/client/index.js [hunk 4]
```diff
@@ -0,0 +1,1710 @@
...
+      } else {
+        brushExtent[1] = new Date(brushExtent[0].getTime() + foucusRangeMS);
+        d3.select('.brush')
+          .call(brush.extent([brushExtent[0], brushExtent[1]]));
+      }
+    }
+
+    var nowDate = new Date(brushExtent[1] - THIRTY_MINS_IN_MS);
+
+    function updateCurrentSGV(entry) {
```
Comment (jasoncalabrese) [inline lib/client/index.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (9)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799258271/issues/source?bid=2452379&fileBranchId=2725007#l337)

File: tests/client.test.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+//    var doc = jsdom('<body></body>');
+//    var window = doc.parentWindow;
+//
+//    global.$ = global.jQuery = require('jquery')(window);
+
+    var serverSettings = {
+      apiEnabled: true, careportalEnabled: true, enabledOptions: "ar2 careportal delta direction upbat errorcodes", defaults: {
+        "units": "mg/dl", "timeFormat": "12", "nightMode": false, "showRawbg": "noise", "customTitle": "Nightscout", "theme": "colors", "alarmUrgentHigh": true, "alarmHigh": true, "alarmLow": true, "alarmUrgentLow": true, "alarmTimeAgoWarn": true, "alarmTimeAgoWarnMins": 15, "alarmTimeAgoUrgent": true, "alarmTimeAgoUrgentMins": 30, "language": "en", "showPlugins": "iob delta direction upbatrawbg"
+      }, "units": "mg/dl", "head": "ae71dca", "version": "0.7.0", "thresholds": {
+        "bg_high": 200, "bg_target_top": 170, "bg_target_bottom": 80, "bg_low": 55
```
Comment (jasoncalabrese) [inline tests/client.test.js]: Codacy Issue found: [Mixed double and single quotes.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799258325/issues/source?bid=2452379&fileBranchId=2725007#l45)

File: lib/client/index.js [hunk 6]
```diff
@@ -0,0 +1,1710 @@
...
+
+      return sel;
+    }
+
+    // if already existing then transition each circle to its new position
+    prepareFocusCircles(focusCircles.transition().duration(UPDATE_TRANS_MS));
+
+    // if new circle then just display
+    prepareFocusCircles(focusCircles.enter().append('circle'))
+      .on('mouseover', function (d) {
```
Comment (jasoncalabrese) [inline lib/client/index.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (10)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799258271/issues/source?bid=2452379&fileBranchId=2725007#l510)

File: tests/client.test.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+  it ('not blow up', function () {
+//    var jsdom = require('jsdom').jsdom;
+//    var doc = jsdom('<body></body>');
+//    var window = doc.parentWindow;
+//
+//    global.$ = global.jQuery = require('jquery')(window);
+
+    var serverSettings = {
+      apiEnabled: true, careportalEnabled: true, enabledOptions: "ar2 careportal delta direction upbat errorcodes", defaults: {
```
Comment (jasoncalabrese) [inline tests/client.test.js]: Codacy Issue found: [Mixed double and single quotes.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799258325/issues/source?bid=2452379&fileBranchId=2725007#l42)

File: tests/client.test.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+  before(function (done) {
+    benv.setup(function() {
+      benv.expose({
+        $: require('jquery')
+        , jQuery: require('jquery')
+        , d3: require('d3')
+        , io: {
+          connect: function mockConnect() {
+            return {
+              on: function mockOn(event, callback) {}
```
Comment (jasoncalabrese) [inline tests/client.test.js]: Codacy Issue found: ['callback' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799258325/issues/source?bid=2452379&fileBranchId=2725007#l18)

File: tests/client.test.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+//    var window = doc.parentWindow;
+//
+//    global.$ = global.jQuery = require('jquery')(window);
+
+    var serverSettings = {
+      apiEnabled: true, careportalEnabled: true, enabledOptions: "ar2 careportal delta direction upbat errorcodes", defaults: {
+        "units": "mg/dl", "timeFormat": "12", "nightMode": false, "showRawbg": "noise", "customTitle": "Nightscout", "theme": "colors", "alarmUrgentHigh": true, "alarmHigh": true, "alarmLow": true, "alarmUrgentLow": true, "alarmTimeAgoWarn": true, "alarmTimeAgoWarnMins": 15, "alarmTimeAgoUrgent": true, "alarmTimeAgoUrgentMins": 30, "language": "en", "showPlugins": "iob delta direction upbatrawbg"
+      }, "units": "mg/dl", "head": "ae71dca", "version": "0.7.0", "thresholds": {
+        "bg_high": 200, "bg_target_top": 170, "bg_target_bottom": 80, "bg_low": 55
+      }, "alarm_types": "predict", "name": "Nightscout"
```
Comment (jasoncalabrese) [inline tests/client.test.js]: Codacy Issue found: [Mixed double and single quotes.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799258325/issues/source?bid=2452379&fileBranchId=2725007#l46)

File: lib/client/index.js [hunk 10]
```diff
@@ -0,0 +1,1710 @@
...
+      socket.emit('ack', currentAlarmType || 'alarm', silenceTime);
+    }
+
+    brushed(false);
+  }
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+//draw a compact visualization of a treatment (carbs, insulin)
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+  function drawTreatment(treatment, scale, showValues) {
```
Comment (jasoncalabrese) [inline lib/client/index.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (12)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799258271/issues/source?bid=2452379&fileBranchId=2725007#l1091)

**Comment by sulkaharo** [comment]
Yay, the best PR! We'll be in so much better shape once this is complete. :)

**Comment by jasoncalabrese** [comment]
This might be a mess for a while...

File: lib/client/index.js (hunk 1 above)
Comment (jasoncalabrese) [inline lib/client/index.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (8)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799298422/issues/source?bid=2452379&fileBranchId=2725007#l1439)

File: lib/client/index.js (hunk 2 above)
Comment (jasoncalabrese) [inline lib/client/index.js]: Codacy Issue found: ['profile' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799298422/issues/source?bid=2452379&fileBranchId=2725007#l8)

File: static/js/ui-utils.js (hunk 3 above)
Comment (jasoncalabrese) [inline static/js/ui-utils.js]: Codacy Issue found: ['rawBGsEnabled' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799298463/issues/source?bid=2452379&fileBranchId=2725007#l5)

File: lib/client/index.js (hunk 4 above)
Comment (jasoncalabrese) [inline lib/client/index.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (9)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799298422/issues/source?bid=2452379&fileBranchId=2725007#l337)

File: lib/client/index.js (hunk 6 above)
Comment (jasoncalabrese) [inline lib/client/index.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (10)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799298422/issues/source?bid=2452379&fileBranchId=2725007#l510)

File: tests/pluginbase.test.js [hunk 11]
```diff
@@ -1,35 +1,53 @@
...
+  before(function (done) {
+    benv.setup(function() {
+      benv.expose({
+        $: require('jquery')
+        , jQuery: require('jquery')
+        , d3: require('d3')
+        , io: {
+          connect: function mockConnect() {
+            return {
+              on: function mockOn(event, callback) {}
```
Comment (jasoncalabrese) [inline tests/pluginbase.test.js]: Codacy Issue found: ['callback' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799298491/issues/source?bid=2452379&fileBranchId=2725007#l17)

File: lib/client/index.js (hunk 10 above)
Comment (jasoncalabrese) [inline lib/client/index.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (12)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799298422/issues/source?bid=2452379&fileBranchId=2725007#l1091)

File: static/js/ui-utils.js [hunk 12]
```diff
@@ -3,119 +3,7 @@
...
-  } catch(err) {
-    console.error(err);
-    showLocalstorageError();
-  }
-
-  return json;
+  return serverSettings.enabledOptions && serverSettings.enabledOptions.indexOf('rawbg') > -1;
 }
 
 function setDefault(variable, defaultValue) {
```
Comment (jasoncalabrese) [inline static/js/ui-utils.js]: Codacy Issue found: ['setDefault' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799298463/issues/source?bid=2452379&fileBranchId=2725007#l9)

File: tests/client.test.js [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+        , bg_target_top: 170
+        , bg_target_bottom: 80
+        , bg_low: 55
+      }
+      , alarm_types: 'predict'
+      , name: 'Nightscout'
+    };
+
+    var plugins = require('../lib/plugins/')().registerClientDefaults();
+    var client = require('../lib/client')(plugins, serverSettings);
```
Comment (jasoncalabrese) [inline tests/client.test.js]: Codacy Issue found: ['client' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799298476/issues/source?bid=2452379&fileBranchId=2725007#l66)

File: lib/client/index.js [hunk 14]
```diff
@@ -0,0 +1,1710 @@
...
+      var message = {'title': 'Last data received ' + [ago.value, ago.label].join(' ')};
+      if (level === 'warn') {
+        generateAlarm(alarmSound, message);
+      } else {
+        generateAlarm(urgentAlarmSound, message);
+      }
+    }
+  }
+
+  function updateTimeAgo() {
```
Comment (jasoncalabrese) [inline lib/client/index.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (11)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799298422/issues/source?bid=2452379&fileBranchId=2725007#l1243)

File: lib/client/index.js (hunk 1 above)
Comment (jasoncalabrese) [inline lib/client/index.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (8)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799312596/issues/source?bid=2452379&fileBranchId=2725007#l1439)

File: lib/client/index.js (hunk 2 above)
Comment (jasoncalabrese) [inline lib/client/index.js]: Codacy Issue found: ['profile' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799312596/issues/source?bid=2452379&fileBranchId=2725007#l8)

File: static/js/ui-utils.js (hunk 3 above)
Comment (jasoncalabrese) [inline static/js/ui-utils.js]: Codacy Issue found: ['rawBGsEnabled' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799312637/issues/source?bid=2452379&fileBranchId=2725007#l5)

File: lib/client/index.js (hunk 4 above)
Comment (jasoncalabrese) [inline lib/client/index.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (9)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799312596/issues/source?bid=2452379&fileBranchId=2725007#l337)

File: lib/client/index.js (hunk 6 above)
Comment (jasoncalabrese) [inline lib/client/index.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (10)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799312596/issues/source?bid=2452379&fileBranchId=2725007#l510)

File: lib/client/index.js (hunk 10 above)
Comment (jasoncalabrese) [inline lib/client/index.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (12)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799312596/issues/source?bid=2452379&fileBranchId=2725007#l1091)

File: static/js/ui-utils.js (hunk 12 above)
Comment (jasoncalabrese) [inline static/js/ui-utils.js]: Codacy Issue found: ['setDefault' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799312637/issues/source?bid=2452379&fileBranchId=2725007#l9)

File: tests/client.test.js (hunk 13 above)
Comment (jasoncalabrese) [inline tests/client.test.js]: Codacy Issue found: ['client' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799312650/issues/source?bid=2452379&fileBranchId=2725007#l66)

File: lib/client/index.js (hunk 14 above)
Comment (jasoncalabrese) [inline lib/client/index.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (11)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799312596/issues/source?bid=2452379&fileBranchId=2725007#l1243)

File: lib/client/index.js [hunk 15]
```diff
@@ -0,0 +1,1710 @@
...
+    if (!brush) {
+      return false;
+    }
+
+    var time = brush.extent()[1].getTime();
+
+    return !alarmingNow() && time - TWENTY_FIVE_MINS_IN_MS < now;
+  }
+
+  function brushed(skipTimer) {
```
Comment (jasoncalabrese) [inline lib/client/index.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (11)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/799312596/issues/source?bid=2452379&fileBranchId=2725007#l310)

**Comment by jasoncalabrese** [comment]
Getting a lot closer, main thing left now is getting test coverage up, but that may cause some more refactoring

**Comment by jasoncalabrese** [comment]
Been working good on my production site, would be nice to merge tomorrow

**Comment by stavlor** [comment]
Its been working well for me as well

**Pull request merged**